bip-signmessage-redone.mediawiki

```diff
@@ -0,0 +1,156 @@
+
+The astute reader will notice that the above is missing a critical part, namely the pubkey <code>kG</code>, without which the verifier cannot actually verify the message. The current message signing standard solves this via a cryptographic trick, wherein the signature <code>S</code> above is a special "recoverable signature" type. Given the message <code>M</code> and the signature <code>S</code>, it is then possible to recover the pubkey <code>kG</code>. The system thus derives the address for the pubkey <code>kG</code>, and if it does not match <code>A</code>, the proof is deemed invalid.
+
+While this is a neat trick, it unnecessarily restricts and complicates the message signing mechanism; for instance, it is currently not possible to sign a message for a P2SH address, because there is no pubkey to recover from the resulting signature.
+
+== Motivation ==
+
+The current message signing standard only works for P2PKH (1...) addresses. By extending it to use a Bitcoin Script based approach, it could be made more generic without causing a too big burden on implementers, who most likely have access to Bitcoin Script interpreters already.
+
+== Specification ==
+
+This BIP follows the specification of BIP-325 challenges and solutions.
+
+Let there be two virtual transactions to_spend and to_sign.
+
+The "to_spend" transaction is:
+
+    nVersion = 0
+    nLockTime = 0
+    vin[0].prevout.hash = 0000...000
+    vin[0].prevout.n = 0xFFFFFFFF
+    vin[0].nSequence = 0
+    vin[0].scriptSig = OP_0 PUSH32[ message_hash ]
+    vin[0].scriptWitness = []
+    vout[0].nValue = 0
+    vout[0].scriptPubKey = message_challenge
+
+where message_hash is a BIP340-tagged hash of the message, i.e. sha256_tag(m), where tag = "BIP????-signed-message", and message_challenge is the to be proven (public) key script.
+
+The "to_sign" transaction is:
+
+    nVersion = 0
+    nLockTime = 0
+    vin[0].prevout.hash = to_spend.txid
+    vin[0].prevout.n = 0
+    vin[0].nSequence = 0
+    vout[0].nValue = 0
+    vout[0].scriptPubKey = message_signature
+
+It may include other inputs, to facilitate a proof of funds.
+
+A message signature is considered valid, inconclusive, or invalid based on whether the to_sign transaction is a valid spend of the to_spend transaction or not, according to the following steps (also see Consensus and standard flags section further down):
+
+1. Valid, if it is a successful spend according to the current consensus rules (sometimes called "policy").
+2. Inconclusive, if it is a successful spend according to consensus rules, but NOT according to policy rules
+3. Invalid, if it is not a successful spend according to consensus rules
+
+A proof is the base64-encoding of the message_signature as is. A validator takes the to be proven pubkey and the proof and transforms it to virtual transactions as described above.
```

Comment:
It seems to me that it should be "to be proven script and the message" rather than "pubkey".

The "to_spend" transaction contains a hash of the message in the input and the script "message_challenge" in the ouptut. If no standard script templates are defined that would receive only one pubkey, then the verifier will need to know the script and how to insert the pubkey there, along with a way to construct witness to spend the script (miniscript could help, although won't cover all possible scripts).

Reply:
Yeah, you're right. Added 'message'.

Reply:
I'm not sure what you mean by the second part though -- I think the spending transaction's witness data should be enough, no? I need to write the reference implementation to verify though.

Reply:
Maybe I do not fully get what data the prover would provide. Do they provide the full script (including the pubkeys) and the witness data to satisfy this script, along with any needed signatures ? Will it be the verifier's job to check that the script matches some template that prover and verifier agree beforehand ?

Reply:
The prover provides the `message_signature`:

1. Prover and verifier establish `message` and `address`
2. Prover generates the transactions, signs the second one, and takes the signature (witness data) and base64 encodes that and gives it to verifier.
3. Verifier base64-decodes, regenerates the two transactions, with the signature in place.

Reply:
To be even more clear, the way this is meant to be done is exactly like the Signet approach, except I'm disallowing scriptSigs and only allowing script witness, where Signet allows both types.

I.e. the signet block signature is the equivalent of the proof by the prover here.

Reply:
Ah, I believe I understand now:

The witness data will include the script, if the `scriptPubKey` in the to_spend transaction is p2sh or p2wsh. And the because the prover and verifier agreed on a certain address beforehand, they agreed on the exact script, too.

Reply:
Yup!